File: Java/Easy/IsBalancedBinaryTree.java
// problems link : https://leetcode.com/problems/balanced-binary-tree/description/
/*
* A tree is considered balanced if A height-balanced binary tree is a binary tree
* in which the depth of the two subtrees of every node never differs by more than one.
*
*
* The approach that I took here is to create another method to calculate the depth of a node,
* then check that every node is balanced recursively.
*/
public boolean isBalanced(TreeNode root) {
	if(root == null)  {
		return true;
	}
	int leftHeight = getHeight(root.left);
	int rightHeight = getHeight(root.right);
	if(Math.abs(leftHeight - rightHeight) > 1)
		return false;

	return isBalanced(root.right) && isBalanced(root.left);
}


public int getHeight(TreeNode node){
	if(node == null)
		return 0;
	final int leftHeight = getHeight(node.left);
	final int rightHeight = getHeight(node.right);

	return Math.max(leftHeight,rightHeight) + 1;
}

File: Kotlin/Easy/IsBalancedBinaryTree.kt
// problems link : https://leetcode.com/problems/balanced-binary-tree/description/
/*
* A tree is considered balanced if A height-balanced binary tree is a binary tree
* in which the depth of the two subtrees of every node never differs by more than one.
*
*
* The approach that I took here is to create another method to calculate the depth of a node,
* then check that every node is balanced recursively.
*/

fun isBalanced(root: TreeNode?): Boolean {
    if (root == null)
        return true
    val leftHand = getHeight(root.left)
    val rightHand = getHeight(root.right)

    if(abs(leftHand - rightHand) > 1)
        return false
    return isBalanced(root.right) && isBalanced(root.left)
}

fun getHeight (root: TreeNode?): Int {
    if(root == null)
        return 0

    val leftHand = getHeight(root.left)
    val rightHand = getHeight(root.right)

    return maxOf(leftHand, rightHand) + 1
}